module/zfs/dbuf.c

```diff
@@ -1096,9 +1110,13 @@ dbuf_read_impl(dmu_buf_impl_t *db, zio_t *zio, uint32_t flags)
 
 	dbuf_add_ref(db, NULL);
 
+	zio_flags = (flags & DB_RF_CANFAIL) ?
+	    ZIO_FLAG_CANFAIL : ZIO_FLAG_MUSTSUCCEED;
+	if ((flags & DB_RF_NO_DECRYPT) && BP_IS_ENCRYPTED(db->db_blkptr))
+		zio_flags |= ZIO_FLAG_RAW;
```

Comment:
> I wonder if we could rename DB_RF_NO_DECRYPT to DB_RF_RAW and do it for non-encrypted BP's too. Because callers NO_DECRYPT callers need to handle compressed bufs. Unless - is DB_RF_NO_DECRYPT only valid for bonus buffers and blocks of dnodes, neither of which are compressed (when encrypted)? If that's the case, let's add assertions to that effect

Yes, `DB_RF_NO_DECRYPT` is currently only used for dnode blocks. I can add assertions for that.

> can we use the more specific ZIO_FLAG_RAW_ENCRYPT here?

`ZIO_FLAG_RAW_ENCRYPT` must always be specified with `ZIO_FLAG_RAW_COMPRESS`. I believe the zio layer asserts this (if not I will make sure to add it). This is consistent with the `ARC_BUF_FLAG`s. The idea is that you should never be attempting to decompress raw encrypted data.